plugins/elantp/fu-elantp-hid-haptic-device.c

```diff
@@ -0,0 +1,865 @@
+fu_elantp_hid_haptic_device_send_cmd(FuDevice *self,
+			      guint8 *tx,
+			      gsize txsz,
+			      guint8 *rx,
+			      gsize rxsz,
+			      GError **error)
+{
+
+	g_autofree guint8 *buf = NULL;
+	gsize bufsz = rxsz + 3;
+
+	if (g_getenv("FWUPD_ELANTP_VERBOSE") != NULL)
+		fu_dump_raw(G_LOG_DOMAIN, "SetReport", tx, txsz);
+
+	if (!fu_udev_device_ioctl(FU_UDEV_DEVICE(self),
+				  HIDIOCSFEATURE(txsz),
+				  tx,
+				  NULL,
+				  FU_ELANTP_DEVICE_IOCTL_TIMEOUT,
+				  error)) 
+		return FALSE;
+
+	if (rxsz == 0)
+		return TRUE;
+
+	/* GetFeature */
+	buf = g_malloc0(bufsz);
+	buf[0] = tx[0]; /* report number */
+	if (!fu_udev_device_ioctl(FU_UDEV_DEVICE(self),
+				  HIDIOCGFEATURE(bufsz),
+				  buf,
+				  NULL,
+				  FU_ELANTP_DEVICE_IOCTL_TIMEOUT,
+				  error))
+		return FALSE;
+	
+	if (g_getenv("FWUPD_ELANTP_VERBOSE") != NULL)
+		fu_dump_raw(G_LOG_DOMAIN, "GetReport", buf, bufsz);
+
+	/* success */
+	return fu_memcpy_safe(rx,
+			      rxsz,
+			      0x0, /* dst */
+			      buf,
+			      bufsz,
+			      0x3, /* src */
+			      rxsz,
+			      error);
+}
+
+static gboolean
+fu_elantp_hid_haptic_device_read_cmd(FuDevice *self,
+			      guint16 reg,
+			      guint8 *rx,
+			      gsize rxsz,
+			      GError **error)
+{
+	guint8 buf[5] = {0x0d, 0x05, 0x03};
+	fu_memwrite_uint16(buf + 0x3, reg, G_LITTLE_ENDIAN);
+	return fu_elantp_hid_haptic_device_send_cmd(self, buf, sizeof(buf), rx, rxsz, error);
+}
+
+static gint
+fu_elantp_hid_haptic_device_write_cmd(FuDevice *self, guint16 reg, guint16 cmd, GError **error)
+{
+	guint8 buf[5] = {0x0d};
+	fu_memwrite_uint16(buf + 0x1, reg, G_LITTLE_ENDIAN);
+	fu_memwrite_uint16(buf + 0x3, cmd, G_LITTLE_ENDIAN);
+	return fu_elantp_hid_haptic_device_send_cmd(self, buf, sizeof(buf), NULL, 0, error);
+}
+
+static gboolean
+fu_elantp_hid_haptic_device_ensure_iap_ctrl(FuDevice *parent, FuElantpHidHapticDevice *self, GError **error)
+{
+	guint8 buf[2] = {0x0};
+	if (!fu_elantp_hid_haptic_device_read_cmd(parent, ETP_CMD_I2C_IAP_CTRL, buf, sizeof(buf), error)) {
+		g_prefix_error(error, "failed to read IAPControl: ");
+		return FALSE;
+	}
+	self->tp_iap_ctrl = fu_memread_uint16(buf, G_LITTLE_ENDIAN);
+
+	/* in bootloader mode? */
+	if ((self->tp_iap_ctrl & ETP_I2C_MAIN_MODE_ON) == 0)
+		fu_device_add_flag(FU_DEVICE(parent), FWUPD_DEVICE_FLAG_IS_BOOTLOADER);
+	else
+		fu_device_remove_flag(FU_DEVICE(parent), FWUPD_DEVICE_FLAG_IS_BOOTLOADER);
+
+	return TRUE;
+}
+
+static gint16
+fu_elantp_hid_haptic_device_ensure_eeprom_iap_ctrl(FuDevice *parent, FuElantpHidHapticDevice *self, GError **error)
+{
+	guint8 buf[2] = {0x0};
+	if (!fu_elantp_hid_haptic_device_read_cmd(parent, 
+						ETP_CMD_I2C_SET_EEPROM_CTRL, 
+						buf, 
+						sizeof(buf), 
+						error)) {
+		g_prefix_error(error, "failed to read IAPControl: ");
+		return -1;
+	}
+	self->iap_ctrl = fu_memread_uint16(buf, G_LITTLE_ENDIAN);
+
+	if ((self->iap_ctrl & 0x800) != 0x800) {
+        	g_prefix_error(error, "bit11 fail: ");
```

Comment:
g_set_error here, g_prefix_error() is only when you have a GError already set and you want to add more info.